doc/developer/design/20230411_faster_dataflow_shutdown.md

```diff
@@ -0,0 +1,255 @@
+
+Failing to shut down a dataflow, or failing to do so in a timely manner, usually has a negative impact on usability:
+
+- Dropped dataflows keep running in the background and compete with useful dataflows for system resources. This can cause noticeably reduced query responsiveness.
+- Dropped dataflows keep ingesting and multiplying data, which might OOM their replica.
+- Dataflows are visible in introspection sources long after they have been dropped, confusing diagnostic queries.
+
+For these reasons, we are interested in completing dataflow shutdowns as quickly as possible.
+
+[`Worker::drop_dataflow`]: https://dev.materialize.com/api/rust/timely/worker/struct.Worker.html#method.drop_dataflow
+[timely-dataflow/#519]: https://github.com/TimelyDataflow/timely-dataflow/pull/519
+[WMR]: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20221204_with_mutually_recursive.md
+[#7577]: https://github.com/MaterializeInc/materialize/issues/7577
+[#16800]: https://github.com/MaterializeInc/materialize/issues/16800
+[#16860]: https://github.com/MaterializeInc/materialize/issues/16860
+
+
+# Explanation
+[explanation]: #explanation
+
+In COMPUTE, every dataflow source holds a reference to a token.
+The COMPUTE worker also holds references to these tokens, associated with the dataflow exports fed by the respective sources.
+When a worker decides that a dataflow export (e.g., and index or an MV) is no longer needed, it drops the tokens associated with it.
+Dataflow sources can observe the dropping of their tokens.
+Once they observe that there is no other holder of their token left, they know to shut down.
+
+In this design we propose three measures to tackle the problems outlined in [motivation], all based on the existing token-based approach:
+
+1. [Fuse operators](#fuse-operators)
+1. [Faster persist source shutdown](#faster-persist-source-shutdown)
+1. [Tokenizing join operators](#tokenizing-join-operators)
+
+Each of these measures introduces the possibility that downstream operators receive incomplete update sets for a given time slice.
+We discuss how to deal with issues introduced by this as well.
+
+## Fuse Operators
+
+We propose the introduction of a "fuse" operator.
+The purpose of this operator is to interrupt the flow of updates when the dataflow is dropped.
+This is useful to ensure downstream operators are not fed more updates when they should be shutting down.
+It is also useful to construct an artificial fixpoint in dataflow loops.
+
+In the scope of this document, the main use case for the fuse operator is forcing divergent WMR dataflows to shut down.
+WMR dataflows circulate updates until they reach a fixpoint, and a WMR dataflow becomes divergent when no fixpoint can ever be reached.
+By introducing a fuse operator in the loop, the circulation of updates is stopped when the dataflow is dropped, allowing the operators in the loop to advance to the empty frontier and shut down.
```

Comment:
I will flesh out this part a bit more, but the gist is:
* Only stopping data updates would be enough to solve the WMR fixpoint issue.
* Also stopping progress updates (by advancing to the empty frontier) potentially speeds up the dataflow shutdown because downstream operators can start shutting down immediately when the dataflow is dropped, instead of having to wait for upstream operators to finish processing their remaining updates.
* As such, advancing to the empty frontier is not a necessity, but we should still do it because it speeds things up and is low-risk.

It sounds like you disagree about the low-risk part. Is this just because of the uncertainty about whether the fuse operator continues to consume upstream data after it has advanced to the empty frontier, or do you worry about something else?

Reply:
I think that my worry is a bit broader, but perhaps motivated only by ignorance of all the details of the written contract for the Timely scheduler. :-)  So trying to articulate it a bit more:

* I agree that stopping data updates indeed is enough to solve the WMR issue. It could be done with a fuse operator or coming to think of it, simply by rendering a `branch_when` that performs a test of the token. So if we do not have the bit of downgrading to the empty frontier, perhaps a specialized fuse operator is just not necessary?
* Let's say that we want to force the downgrade to the empty frontier in the fuse operator. Now, to ensure the design is sound, we need a correctness argument that this downgrading is safe under the Timely scheduling constraints for any arbitrarily shaped dataflow graph, with as many loops as you want it to have. Perhaps I am just missing something, but I could not grok this from the design doc. I was hoping to get some pointers to (a) the guarantees that we are using regarding Timely scheduling, backed by the Timely documentation; (b) the reasoning that gets me the implication from these guarantees to the safety of deploying the empty-frontier forcing fuse.
* I did not look at the details, but it seems that after the empty-frontier forcing change to fuse, we have still some pending problems in the fast `sqllogictest` run where we are getting stuck in a WMR query: https://buildkite.com/materialize/tests/builds/53595#01877af5-2a11-43b0-846d-8952679c4386. Not sure if it is a simple bug or if something bigger is lurking behind it. It could be that we need to drop the capability there, not only downgrade it to the empty frontier, but I am not 100% sure.
* I tried to think a bit about how the empty-frontier forcing fuse behaves in terms of a path summary, but I am still a bit uncertain about the semantics. Do you have a handy argument for why it works out? I will keep on thinking about a justification along these lines.

Reply:
Thanks for writing your thoughts down! I don't have answer for all your points yet, but these are my thoughts to some of them:

> So if we do not have the bit of downgrading to the empty frontier, perhaps a specialized fuse operator is just not necessary?

For me adding a token check to the WMR rendering is the specialization and having a fuse operator would be the generic implementation. The fuse operator would be useful in other contexts as well, and it wouldn't be less useful in the WMR context, which is why I'd prefer that solution even if it doesn't actively downgrade its frontier on token drop.

> Now, to ensure the design is sound, we need a correctness argument that this downgrading is safe under the Timely scheduling constraints for any arbitrarily shaped dataflow graph, with as many loops as you want it to have.

Can we define what "safe" means here? Is it just that all operators in the dataflow will shut down eventually and free all their resources or is it more?

> I did not look at the details, but it seems that after the empty-frontier forcing change to fuse, we have still some pending problems in the fast sqllogictest run where we are getting stuck in a WMR query

Good catch, I'll need to look at that.

> I tried to think a bit about how the empty-frontier forcing fuse behaves in terms of a path summary, but I am still a bit uncertain about the semantics. Do you have a handy argument for why it works out?

I will admit that I don't fully grok that either. My understanding is that setting the output path summary to the empty antichain makes it look to the reachability checker as if there is no connection between the fuse operator and its downstream node. I guess this only applies to reachability propagation from other nodes though, because the downstream node is still able to see the frontier of the fuse operator in its input frontier. When the fuse operator drops its capability, then that advances its output frontier to the empty frontier, even though upstream nodes still hold capabilities that would hold back the frontier if the path summary wasn't the empty antichain.

I'd be fine with moving forward with a fuse operator that leaves its output frontier alone. It would still be great if we could answer your questions, but I don't think we need to block on them.

Reply:
Regarding the safety question: I think that eventual shutdown and no resource leakage are a fine start, but we should also consider other aspects, such as things not getting stuck or data exchanged not violating the frontiers communicated. I think there is a bigger discussion in this [paper](https://www21.in.tum.de/~traytel/papers/itp21-progress_tracking/safe.pdf). This might be a good candidate for a reading group discussion if Frank is available and we'd like to go into the details.

Reply:
I decided that it's fine to have a fuse operator that doesn't do special things to its output frontier, especially since #18718 now merged without this feature. I adjusted the design doc to mark this as a possible future optimization we can do.